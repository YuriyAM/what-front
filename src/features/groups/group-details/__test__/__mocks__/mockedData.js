export const groupDetailsMock = {
  id: 1,
  studentGroupData: {
    data: {
      courseId:1,
      finishDate: "2021-12-24T00:00:00",
      id: 1,
      mentorIds:[ 6 ],
      name: "Soft Skills for Lecturers - 2021/2",
      startDate: "2021-07-05T00:00:00",
      studentIds: [ 2, 3, 16, 17, 18, 4 ]
    },
    isLoading: false,
    isLoaded: true,
    error: ''
  },
  mentorsData: {
    data: [
      {
        id: 1,
        email: "michael.taylor@example.com",
        firstName: "Michael",
        lastName: "Taylor",
        avatarUrl: "https://whatapistorageacc.blob.core.windows.net/28edbc17f2394a5c967e5f5568bc523a/7.jpg"
      },
      {
        id: 2,
        email: "william.davies@example.com",
        firstName: "Williames",
        lastName: "Daviess",
        avatarUrl: "https://whatapistorageacc.blob.core.windows.net/bd18d83832094662877fdd44ded8ca63/9.jpg"
      },
      {
        id: 3,
        email: "david.brown@example.com",
        firstName: "Dav",
        lastName: "Brow",
        avatarUrl: "https://whatapistorageacc.blob.core.windows.net/32f19972bebb42a689104071a11c5e33/11.jpg"
      },
      {
        id: 4,
        email: "richard.thomas@example.com",
        firstName: "Richard",
        lastName: "Thomas",
        avatarUrl: "https://whatapistorageacc.blob.core.windows.net/fbbfa551fede45f5940ca750a3f7986b/13.jpg"
      },
      {
        id: 5,
        email: "joseph.evans@example.com",
        firstName: "Joseph",
        lastName: "Evans",
        avatarUrl: "https://whatapistorageacc.blob.core.windows.net/b1a46fee2f8c4fdbbffa53465641654c/15.jpg"
      },
      {
        id: 6,
        email: "linda.jackson@example.com",
        firstName: "Linda",
        lastName: "Jackson",
        avatarUrl: "https://whatapistorageacc.blob.core.windows.net/1d9a34a8b9494a57b5bf87fe87a14c1b/8.jpg"
      },
      {
        id: 7,
        email: "elizabeth.hill@example.com",
        firstName: "Elizabeth",
        lastName: "Hill",
        avatarUrl: "https://whatapistorageacc.blob.core.windows.net/e80c2bd3e48a4d8b8eeb95d41fcb6e52/10.jpg"
      },
      {
        id: 8,
        email: "barbara.harris@example.com",
        firstName: "Barbara",
        lastName: "Harris",
        avatarUrl: "https://whatapistorageacc.blob.core.windows.net/b2c214175eb74adabecd2b9747d8dc85/12.jpg"
      },
      {
        id: 9,
        email: "susan.clark@example.com",
        firstName: "Susan",
        lastName: "Clark",
        avatarUrl: "https://whatapistorageacc.blob.core.windows.net/630846a66d874499b529cb5ad4db82e2/14.jpg"
      },
      {
        id: 10,
        email: "jessica.cooper@example.com",
        firstName: "Jessica",
        lastName: "Cooper",
        avatarUrl: "https://whatapistorageacc.blob.core.windows.net/1736a6ec4a4845349b38033b613982c9/16.jpg"
      },
      {
        id: 11,
        email: "fhfhg@mail.com",
        firstName: "sertered",
        lastName: "gfhgjfh",
        avatarUrl: null
      },
      {
        id: 13,
        email: "isc06288@zwoho.com",
        firstName: "Second",
        lastName: "User",
        avatarUrl: null
      },
      {
        id: 14,
        email: "qwe@qwe.com",
        firstName: "fgjjd",
        lastName: "jdfgjfdgj",
        avatarUrl: null
      }
    ],
    isLoading: false,
    isLoaded: true,
    error: ''
  },
  coursesData: {
    data: [
      {
        id: 1,
        name: "Soft Skills for Lecturers",
        isActive: true
      },
      {
        id: 2,
        name: "C# Programming",
        isActive: true
      },
      {
        id: 3,
        name: "3D Modelling",
        isActive: true
      },
      {
        id: 4,
        name: "Cybersecurity",
        isActive: true
      },
      {
        id: 5,
        name: "Advanced Technical English",
        isActive: true
      },
      {
        id: 6,
        name: "Intermediate Technical English",
        isActive: true
      },
      {
        id: 7,
        name: "course1",
        isActive: true
      },
      {
        id: 8,
        name: "NewCoolCours",
        isActive: true
      },
      {
        id: 9,
        name: "course28фы",
        isActive: true
      },
      {
        id: 10,
        name: "course4",
        isActive: true
      },
      {
        id: 11,
        name: "course5",
        isActive: true
      },
      {
        id: 12,
        name: "course6",
        isActive: true
      },
      {
        id: 13,
        name: "English for IT",
        isActive: true
      }
    ],
    isLoading: false,
    loaded: true,
    error: ''
  },
  studentsData: {
    data: [
      {
      id: 1,
      email: "david.brown@example.com",
      firstName: "Dav",
      lastName: "Brow",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/32f19972bebb42a689104071a11c5e33/11.jpg"
    },
    {
      id: 2,
      email: "richard.thomas@example.com",
      firstName: "Richard",
      lastName: "Thomas",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/fbbfa551fede45f5940ca750a3f7986b/13.jpg"
    },
    {
      id: 3,
      email: "joseph.evans@example.com",
      firstName: "Joseph",
      lastName: "Evans",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/b1a46fee2f8c4fdbbffa53465641654c/15.jpg"
    },
    {
      id: 4,
      email: "thomas.roberts@example.com",
      firstName: "Thomas",
      lastName: "Roberts",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/ade5fa3adb51431e848e0355cbaaec6c/17.jpg"
    },
    {
      id: 5,
      email: "charles.johnson@example.com",
      firstName: "Charles",
      lastName: "Johnson",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/4bc338e6e98d489dbf170d719c74fe68/19.jpg"
    },
    {
      id: 6,
      email: "christopher.wilson@example.com",
      firstName: "Christopher",
      lastName: "Wilson",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/4b5d7bf862b24ed0aaad07b5c916de0f/21.jpg"
    },
    {
      id: 7,
      email: "daniel.robinson@example.com",
      firstName: "Daniel",
      lastName: "Robinson",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/f742931ff3f74ff59efdee9433d3ac41/23.jpg"
    },
    {
      id: 8,
      email: "matthew.wright@example.com",
      firstName: "Matthew",
      lastName: "Wright",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/cf512d51ac6b4cd982b3e5039aea664f/25.jpg"
    },
    {
      id: 9,
      email: "anthony.wood@example.com",
      firstName: "Anthony",
      lastName: "Wood",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/4a607d89efe4460580f694e10e23314d/27.jpg"
    },
    {
      id: 10,
      email: "mark.thompson@example.com",
      firstName: "Mark",
      lastName: "Thompson",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/7646d6d322434e8cbcd29fbf6dd030d2/29.jpg"
    },
    {
      id: 11,
      email: "donald.hall@example.com",
      firstName: "Donald",
      lastName: "Hall",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/9ee0c7d8e01346ee844ae2893d7e4df7/31.jpg"
    },
    {
      id: 12,
      email: "steven.green@example.com",
      firstName: "Steven",
      lastName: "Green",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/e7de05778f1e4fd7b208451a94cc56d2/33.jpg"
    },
    {
      id: 13,
      email: "paul.walker@example.com",
      firstName: "Paul",
      lastName: "Walker",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/feeaecdd6aa347dfa5d4b0ddda771ec5/35.jpg"
    },
    {
      id: 14,
      email: "andrew.hughes@example.com",
      firstName: "Andrew",
      lastName: "Hughes",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/5680a096834f40f6b7ad6292151657bc/37.jpg"
    },
    {
      id: 15,
      email: "joshua.edwards@example.com",
      firstName: "Joshua",
      lastName: "Edwards",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/5a7bfffeeb9d47d68c1599eb68e32eae/39.jpg"
    },
    {
      id: 16,
      email: "barbara.harris@example.com",
      firstName: "Barbara",
      lastName: "Harris",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/b2c214175eb74adabecd2b9747d8dc85/12.jpg"
    },
    {
      id: 17,
      email: "susan.clark@example.com",
      firstName: "Susan",
      lastName: "Clark",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/630846a66d874499b529cb5ad4db82e2/14.jpg"
    },
    {
      id: 18,
      email: "jessica.cooper@example.com",
      firstName: "Jessica",
      lastName: "Cooper",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/1736a6ec4a4845349b38033b613982c9/16.jpg"
    },
    {
      id: 19,
      email: "sarah.harrison@example.com",
      firstName: "Sarah",
      lastName: "Harrison",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/8c1a55663a94456ca3f6c58749bb90a7/18.jpg"
    },
    {
      id: 20,
      email: "karen.ward@example.com",
      firstName: "Karen",
      lastName: "Ward",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/99e8d584c6af4ff793c13b174202cfd6/20.jpg"
    },
    {
      id: 21,
      email: "nancy.martin@example.com",
      firstName: "Nancy",
      lastName: "Martin",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/17f4f623739b4adcb6d4e5fd1bae377c/22.jpg"
    },
    {
      id: 22,
      email: "lisa.davis@example.com",
      firstName: "Lisa",
      lastName: "Davis",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/45873fc69dac423db15c04d9ddadee2b/24.jpg"
    },
    {
      id: 23,
      email: "betty.baker@example.com",
      firstName: "Betty",
      lastName: "Baker",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/bf5f8faa55ad4faf947b2c7a5b636293/26.jpg"
    },
    {
      id: 24,
      email: "margaret.morris@example.com",
      firstName: "Margaret",
      lastName: "Morris",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/5eb6b24a5e054e169c4e7074444ec950/28.jpg"
    },
    {
      id: 25,
      email: "sandra.james@example.com",
      firstName: "Sandra",
      lastName: "James",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/d8e771dec4ab484abf5478285a822d37/30.jpg"
    },
    {
      id: 26,
      email: "ashley.king@example.com",
      firstName: "Ashley",
      lastName: "King",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/672f926896aa4cba996f68c235ddd2d4/32.jpg"
    },
    {
      id: 27,
      email: "kimberly.morgan@example.com",
      firstName: "Kimberly",
      lastName: "Morgan",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/5558128d1c79495a934793bd0d63b2e2/34.jpg"
    },
    {
      id: 28,
      email: "emily.allen@example.com",
      firstName: "Emily",
      lastName: "Allen",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/30c4b294f96e41b5b29b3709deb0b3e0/36.jpg"
    },
    {
      id: 29,
      email: "donna.moore@example.com",
      firstName: "Donna",
      lastName: "Moore",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/70a21d6d5fb4467c81b812dc89ec3e58/38.jpg"
    },
    {
      id: 30,
      email: "michelle.parker@example.com",
      firstName: "Michelle",
      lastName: "Parker",
      avatarUrl: "https://whatapistorageacc.blob.core.windows.net/db23452c773a4ec6a429f91538841d7a/40.jpg"
    },
    {
      id: 31,
      email: "ilya@gmail.comn",
      firstName: "iLYA",
      lastName: "ILYA",
      avatarUrl: null
    },
    {
      id: 33,
      email: "",
      firstName: "",
      lastName: "Donna Moore",
      avatarUrl: null
    },
    {
      id: 34,
      email: "qweqwe@wjsdf.sgdg",
      firstName: "Qweqweqwe",
      lastName: "Qweqwe",
      avatarUrl: null
    }
  ],
    isLoading: false,
    isLoaded: true,
    error: ''
  }
}
