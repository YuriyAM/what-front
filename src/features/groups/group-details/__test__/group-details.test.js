import React from 'react';
import { Router } from "react-router-dom";
import { useSelector } from 'react-redux';
import { cleanup, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event'
import { createMemoryHistory } from 'history';

import { paths } from '@/shared';
import { commonHelpers } from "@/utils";
import { GroupDetails } from '../group-details';
import { groupDetailsMock } from "./__mocks__/mockedData.js";

jest.mock("react-redux", () => ({ useSelector: jest.fn() }));
jest.mock("@/shared/hooks", () => ({ useActions: jest.fn() }));
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn().mockReturnValue({ push: jest.fn() }),
}));

describe('Render of GroupDetails', () => {
  let studentGroupSelector;
  let mentorsSelector;
  let coursesSelector;
  let studentsSelector;
  let historyMock;
  let id;

  beforeEach(() => {
    studentGroupSelector = groupDetailsMock.studentGroupData;
    mentorsSelector = groupDetailsMock.mentorsData;
    coursesSelector = groupDetailsMock.coursesData;
    studentsSelector = groupDetailsMock.studentsData;
    id = groupDetailsMock.id;

    historyMock = { push: jest.fn(), location: {}, listen: jest.fn(), createHref: jest.fn() };
    commonHelpers.transformDateTime = jest.fn().mockReturnValue({ formInitialValue: '2022-06-16' });

    useSelector
      .mockReturnValueOnce(studentGroupSelector)
      .mockReturnValueOnce(mentorsSelector)
      .mockReturnValueOnce(coursesSelector)
      .mockReturnValueOnce(studentsSelector);
  });
  
  afterEach(cleanup);

  it('Should the component be rendered.', () => {

    const { container } = render(
      <Router history={historyMock}>
        <GroupDetails id={id}
                      studentGroupData={studentGroupSelector}
                      studentsData={studentsSelector}
                      mentorsData={mentorsSelector}
                      coursesData={coursesSelector} />
      </Router>
    );

    const groupDetailsContainer = container.getElementsByClassName('container');

    expect(groupDetailsContainer).toMatchSnapshot();
  });

  it('should loader appear when data is false', () => {
    studentGroupSelector = {
      data: groupDetailsMock.studentGroupData.data,
      isLoading: true,
      isLoaded: false,
      error: ''
    };

    useSelector.mockReturnValue(studentGroupSelector);

    const { container } = render(
      <Router history={historyMock}>
        <GroupDetails id={id}
          studentGroupData={studentGroupSelector}
          studentsData={studentsSelector}
          mentorsData={mentorsSelector}
          coursesData={coursesSelector} />
      </Router>);

    const loader = container.querySelector('.spinner-border');
    expect(loader).toBeInTheDocument();
  });

  it('should the h2 element contain "Group: Soft Skills for Lecturers - 2021/2"', () => {
    const { container } = render(
      <Router history={historyMock}>
        <GroupDetails id={id}
          studentGroupData={studentGroupSelector}
          studentsData={studentsSelector}
          mentorsData={mentorsSelector}
          coursesData={coursesSelector} />
      </Router>);

    const header = container.querySelector('h2');
    expect(header.innerHTML).toBe('Group: Soft Skills for Lecturers - 2021/2');
  });

  it('should the table content match', () => {
    const { container } = render(
      <Router history={historyMock}>
        <GroupDetails id={id}
          studentGroupData={studentGroupSelector}
          studentsData={studentsSelector}
          mentorsData={mentorsSelector}
          coursesData={coursesSelector} />
      </Router>);

    const table = container.querySelector('tbody');
    expect(table.innerHTML).toBe('<tr><td>1</td><td>Richard Thomas</td><td>richard.thomas@example.com</td></tr><tr><td>2</td><td>Joseph Evans</td><td>joseph.evans@example.com</td></tr><tr><td>3</td><td>Thomas Roberts</td><td>thomas.roberts@example.com</td></tr><tr><td>4</td><td>Barbara Harris</td><td>barbara.harris@example.com</td></tr><tr><td>5</td><td>Susan Clark</td><td>susan.clark@example.com</td></tr><tr><td>6</td><td>Jessica Cooper</td><td>jessica.cooper@example.com</td></tr>');
  });

  it('should url change to "/students/2"', () => {
    const { container } = render(
      <Router history={historyMock}>
        <GroupDetails id={id}
          studentGroupData={studentGroupSelector}
          studentsData={studentsSelector}
          mentorsData={mentorsSelector}
          coursesData={coursesSelector} />
      </Router>);


    userEvent.click(screen.getByText('Richard Thomas'));
    window.history.pushState({}, 'Richard Thomas page', '/students/2');
    expect(global.window.location.pathname).toEqual('/students/2');
  });

  it('Should redirect to 404 page in case of error in course details data.', () => {
    const history = createMemoryHistory();

    useSelector.mockReturnValue({
      ...studentGroupSelector,
      error: 'Something went wrong',
      loaded: false,
    });

    render(
      <Router history={history}>
         <GroupDetails id={id}
          studentGroupData={studentGroupSelector}
          studentsData={studentsSelector}
          mentorsData={mentorsSelector}
          coursesData={coursesSelector} />
      </Router>
    );

    history.push(paths.NOT_FOUND);
    expect(history.location.pathname).toBe(paths.NOT_FOUND);
  });

});