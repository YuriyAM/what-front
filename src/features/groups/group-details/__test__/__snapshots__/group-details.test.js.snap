// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render of GroupDetails Should the component be rendered. 1`] = `
HTMLCollection [
  <div
    class="container pt-5"
  >
    <div
      class="row justify-content-center"
    >
      <div
        class="w-100 card shadow p-4"
      >
        <div
          class="d-flex flex-row text-left justify-content-between"
        >
          <div
            class="d-flex flex-column"
          >
            <h2>
              Group: 
              Soft Skills for Lecturers - 2021/2
            </h2>
            <p
              class="m-0"
            >
               - 
            </p>
          </div>
          <div
            class="pt-3"
          >
            <a>
              <span>
                View schedule
              </span>
            </a>
          </div>
        </div>
        <hr
          class="p-0"
        />
        <div
          class="d-flex mb-2"
        >
          <h4
            class="pr-2 mb-2"
          >
            Mentors:
          </h4>
          <div
            class="d-flex flex-wrap"
          >
            <div
              class="pr-2 lead"
            >
              <span
                class="badge badge-pill badge-info"
              >
                <a
                  class="text-decoration-none text-light"
                >
                  Linda
                   
                  Jackson
                </a>
              </span>
            </div>
          </div>
        </div>
        <div
          class="d-flex align-items-center mb-2 lead"
        >
          <h4
            class="mb-2 pr-4"
          >
            Course:
          </h4>
          <span
            class="badge badge-pill badge-info"
          >
            <a
              class="text-decoration-none text-white"
            >
              Soft Skills for Lecturers
            </a>
          </span>
        </div>
        <h4
          class="h4 my-2"
        >
          Students:
        </h4>
        <div
          class="table-responsive"
        >
          <table
            class="table table-bordered table-hover"
          >
            <thead>
              <tr>
                <th>
                  №
                </th>
                <th>
                  Name
                </th>
                <th>
                  Email
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  1
                </td>
                <td>
                  Richard
                   
                  Thomas
                </td>
                <td>
                  richard.thomas@example.com
                </td>
              </tr>
              <tr>
                <td>
                  2
                </td>
                <td>
                  Joseph
                   
                  Evans
                </td>
                <td>
                  joseph.evans@example.com
                </td>
              </tr>
              <tr>
                <td>
                  3
                </td>
                <td>
                  Thomas
                   
                  Roberts
                </td>
                <td>
                  thomas.roberts@example.com
                </td>
              </tr>
              <tr>
                <td>
                  4
                </td>
                <td>
                  Barbara
                   
                  Harris
                </td>
                <td>
                  barbara.harris@example.com
                </td>
              </tr>
              <tr>
                <td>
                  5
                </td>
                <td>
                  Susan
                   
                  Clark
                </td>
                <td>
                  susan.clark@example.com
                </td>
              </tr>
              <tr>
                <td>
                  6
                </td>
                <td>
                  Jessica
                   
                  Cooper
                </td>
                <td>
                  jessica.cooper@example.com
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>,
]
`;
